src/ray/common/memory_monitor.cc

```diff
@@ -0,0 +1,185 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/memory_monitor.h"
+
+#include <fstream>  // std::ifstream
+#include <tuple>
+
+#include "ray/common/nullable_uint.h"
+#include "ray/common/ray_config.h"
+#include "ray/util/logging.h"
+
+namespace ray {
+
+MemoryMonitor::MemoryMonitor(float usage_threshold_,
+                             uint64_t monitor_interval_ms,
+                             MemoryUsageRefreshCallback monitor_callback)
+    : usage_threshold_(usage_threshold_),
+      monitor_callback_(monitor_callback),
+      io_context_(),
+      monitor_thread_([this] {
+        boost::asio::io_service::work io_service_work_(io_context_);
+        io_context_.run();
+      }),
```

Comment:
What's the consideration of starting a new thread vs passing in an io_context and run there (e.g. raylet already has event loops)?

Reply:
@scv119  thoughts? I tried to stick to the same pattern in file system monitor

Reply:
looks good to me!